### PR TITLE
feat: use UltraJSON if available

### PR DESCRIPTION
**Pull request by jjaakola-aiven**
# About this change - What it does

UltraJson gives slight performance improvement.

# Why this way

Decided to support the vanilla `json` and `ultrajson` for now.

### CODE REVIEW

File: tests/unit/test_avro_compatibility.py [hunk 1]
```diff
@@ -33,8 +33,9 @@
 )
 schema8 = parse_avro_schema_definition(
     '{"type":"record","name":"myrecord","fields":[{"type":"string","name":"f1"},{"type":"string",'
-    '"name":"f2","default":"foo"}]},{"type":"string","name":"f3","default":"bar"}]}'
```
Comment (jjaakola-aiven) [inline tests/unit/test_avro_compatibility.py:36]: This data was broken so I took the trailing data out.

File: tests/integration/test_schema.py [hunk 2]
```diff
@@ -2882,7 +2884,7 @@ async def test_invalid_schema_should_provide_good_error_messages(registry_async_
         f"subjects/{test_subject}/versions",
         json={"schema": schema_str[:-1]},
     )
-    assert res.json()["message"] == "Invalid AVRO schema. Error: Expecting ',' delimiter: line 1 column 18 (char 17)"
+    assert res.json()["message"].startswith("Invalid AVRO schema. Error: ")
```
Comment (jjaakola-aiven) [inline tests/integration/test_schema.py:2887]: The actual error message is different in `ujson` than in `json`.

File: karapace/schema_models.py [hunk 3]
```diff
@@ -37,15 +36,8 @@ def parse_avro_schema_definition(s: str, validate_enum_symbols: bool = True, val
...
     again.
     """
-    try:
-        json_data = json.loads(s)
-    except json.JSONDecodeError as e:
-        if e.msg != "Extra data":
-            raise
-
-        json_data = json.loads(s[: e.pos])
-
```
Comment (jjaakola-aiven) [inline karapace/schema_models.py:47]: I removed this whole block as this accepts invalid JSON. If there is trailing data the data shall be considered bad.
For example `{"type": "fish"},["trailing"]` was splitted with the position from `json.JSONDecodeError` to `{"type": "fish"}`. The result is valid JSON, but should not be accepted as the input is garbage.

**Comment by Fleshgrinder** [comment]
Why not [orjson](https://pypi.org/project/orjson/)? It is significantly faster than all other options, handles more types natively, and codes directly to `bytes`. Thus removing the unnecessary `str` allocation and `str` to `bytes` conversion for sending things over the wire. It also still supports Python 3.7.

**Comment by jjaakola-aiven** [comment]
The `orjson` can be added later if needed. `ujson` is quite straightforward replacement for vanilla `json`.

File: karapace/utils.py [hunk 4]
```diff
@@ -12,14 +12,25 @@
...
+from typing import IO, NoReturn, Optional, overload, Union
 
-import json
+import importlib
 import kafka.client_async
 import logging
 import time
 
+ultrajson = importlib.util.find_spec("ujson")
+if ultrajson:
```
Comment (Fleshgrinder) [inline karapace/utils.py:25]: There is no need for the global variable, is there?

```suggestion
if importlib.util.find_spec("ujson"):
```

File: karapace/utils.py [hunk 5]
```diff
@@ -73,17 +84,34 @@ def default_json_serialization(
...
+    if indent is not None:
+        arguments["indent"] = indent
     res = json.dumps(
         obj,
-        sort_keys=sort_keys,
-        default=default_json_serialization,
-        separators=separators,
+        **arguments,
     )
     return res.encode("utf-8") if binary else res
```
Comment (Fleshgrinder) [inline karapace/utils.py:100]: Using overloads allows us to help editors in understanding what this function is going to return, as it either returns `bytes` or `str` in a predictable way. I also removed the global `SEPARATORS` constant, as there is no need to occupy the memory for it, as there is no reuse. Using `Optional[T] = None` for all arguments makes it possible to write wrapper functions, without copying the actual default values. JSON should also be compact by default, no need to waste precious bytes.

```python
@overload
def json_encode(
    it: Any,
    *,
    binary: Literal[False, None] = None,
    sort_keys: Optional[bool] = None,
    compact: Optional[bool] = None,
    indent: Optional[int] = None,
) -> str:
    ...


@overload
def json_encode(
    it: Any,
    *,
    binary: Literal[True],
    sort_keys: Optional[bool] = None,
    compact: Optional[bool] = None,
    indent: Optional[int] = None,
) -> bytes:
    ...


def json_encode(
    it: Any,
    *,
    binary: Optional[bool] = None,
    sort_keys: Optional[bool] = None,
    compact: Optional[bool] = None,
    indent: Optional[int] = None,
) -> Union[bytes, str]:
    kwargs = {}
    if indent is not None:
        kwargs["indent"] = indent
    if compact is not False and indent is None:  # compact by default
        kwargs["separators"] = (",", ":")
    if sort_keys is True:
        kwargs["sort_keys"] = True
    result = json.dumps(it, default=default_json_serialization, **kwargs)
    return result if binary is True else result.encode("utf8")
```

```python
string = json_encode({})  # correctly recognized as `str`
also_string = json_encode({}, binary=False)  # correctly recognized as `str`
binary = json_encode({}, binary=True)  # correctly recognized as `bytes`
```

File: karapace/utils.py [hunk 6]
```diff
@@ -73,17 +84,34 @@ def default_json_serialization(
...
+def json_decode(content: [str, IO]) -> JsonData:
+    if isinstance(content, str):
+        try:
+            return json.loads(content)
+        except JSONDecodeError as ex:
+            raise ex
+    try:
+        return json.load(content)
+    except JSONDecodeError as ex:
+        raise ex
```
Comment (Fleshgrinder) [inline karapace/utils.py:112]: There is no need to catch and rethrow the same exception, and `bytes` is a perfectly valid input for both `json` and `usjon` as well. By including `bytes` we can also get rid of **ALL** `.decode("utf8")` that are currently sprinkled across the codebase when calling this function.

```suggestion
def json_decode(it: Union[bytes, str, IO]) -> JsonData:
    if isinstance(content, bytes) or isinstance(content, str):
        return json.loads(content)
    else:
        return json.load(content)
```

File: karapace/schema_reader.py [hunk 7]
```diff
@@ -326,8 +325,8 @@ def handle_messages(self) -> None:
                 value = None
                 if msg.value:
                     try:
-                        value = json.loads(msg.value.decode("utf8"))
-                    except json.JSONDecodeError:
+                        value = json_decode(msg.value.decode("utf8"))
```
Comment (Fleshgrinder) [inline karapace/schema_reader.py:328]: Just one of many places of unnecessary binary decoding, if the signature of `json_decode` is updated as suggested in my other comment.

```suggestion
                        value = json_decode(msg.value)
```

**Review by Fleshgrinder — CHANGES_REQUESTED** [review]

**Comment by jjaakola-aiven** [comment]
As this also impacts the formatting of the key I registered two schemas. Another with code from current main and another with this branch. Result of keys:
```
THIS BRANCH -> {"keytype":"SCHEMA","subject":"test-key","version":1,"magic":1}	
MAIN BRANCH -> {"keytype":"SCHEMA","subject":"test-diff","version":1,"magic":1}
```

Both have the same order of fields and are in the most compact form.

File: karapace/utils.py (hunk 4 above)
Comment (jjaakola-aiven) [reply karapace/utils.py]: Done.

File: karapace/utils.py (hunk 5 above)
Comment (jjaakola-aiven) [reply karapace/utils.py]: Done.

File: karapace/utils.py (hunk 6 above)
Comment (jjaakola-aiven) [reply karapace/utils.py]: Yes, rethrow was a workaround to flake8 and pylint unused import nag.

File: karapace/schema_reader.py (hunk 7 above)
Comment (jjaakola-aiven) [reply karapace/schema_reader.py]: Removed the unwanted decoding calls.

File: karapace/utils.py [hunk 8]
```diff
@@ -70,18 +80,53 @@ def default_json_serialization(
...
+    if compact is not False and indent is None:
+        kwargs["separators"] = (",", ":")
+    if sort_keys is True:
+        kwargs["sort_keys"] = True
+    result = json.dumps(obj, default=default_json_serialization, **kwargs)
+    return result.encode("utf8") if binary is True else result
+
+
+def json_decode(content: [AnyStr, IO[AnyStr]]) -> JsonData:
+    if isinstance(content, str) or isinstance(content, bytes):  # pylint: disable=consider-merging-isinstance
```
Comment (Fleshgrinder) [inline karapace/utils.py:127]: Why would we not follow pylint's suggestion? It is more efficient to perform the work in C.

```suggestion
    if isinstance(content, (str, bytes)):
```

**Review by Fleshgrinder — CHANGES_REQUESTED** [review]

File: karapace/utils.py (hunk 8 above)
Comment (Fleshgrinder) [reply karapace/utils.py:127]: Not possible in Python 3.7

**Review by Fleshgrinder — APPROVED** [review]

**Pull request merged**